File: interviews/leetcode/sliding-window/1493. Longest Subarray of Ones After Delete/solution.py
class Solution:
    def longestSubarray(self, nums: List[int]) -> int:

        left = 0
        max_length = 0
        zero_count = 0

        # Sliding window approach
        for right in range(len(nums)):
            if nums[right] == 0:
                zero_count += 1

            # If there are more than one '0's, shrink the window
            while zero_count > 1:
                if nums[left] == 0:
                    zero_count -= 1
                left += 1

            # Calculate the maximum length of the subarray with at most one '0'
            max_length = max(max_length, right - left + 1)

        # If the entire array consists of 1's, the result should be n-1 (since we must delete one element)
        return max_length - 1 if max_length == len(nums) else max_length

File: interviews/leetcode/sliding-window/1493. Longest Subarray of Ones After Delete/solution.go
package main

func longestSubarray(nums []int) int {
	left := 0
	maxLength := 0
	zeroCount := 0

	// Sliding window approach
	for right := 0; right < len(nums); right++ {
		if nums[right] == 0 {
			zeroCount++
		}

		// If there are more than one '0's, move the left pointer to shrink the window
		for zeroCount > 1 {
			if nums[left] == 0 {
				zeroCount--
			}
			left++
		}

		// Calculate the maximum length of subarray with at most one '0'
		maxLength = max(maxLength, right-left) // Update maxLength - we don't add +1 here since the window is already shrunk
	}

	return maxLength
}

/*
func main() {
    nums := []int{1, 1, 0, 1, 1, 0, 1, 1, 1}
    fmt.Println(longestSubarray(nums)) // Output: 6
}
*/
